include/zephyr/net/lwm2m.h

```diff
@@ -1410,7 +1410,23 @@ void lwm2m_rd_client_update(void);
 char *lwm2m_path_log_buf(char *buf, struct lwm2m_obj_path *path);
 
 /** 
- * @brief LwM2M SEND operation to given path list
+ * @brief LwM2M SEND operation to given path list asynchronously with confirmation callback
+ *
+ * @param ctx LwM2M context
+ * @param path_list LwM2M Path string list
+ * @param path_list_size Length of path list. Max size is CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE
+ * @param reply_cb Callback triggered with confirmation state or NULL if not used
+ * 
+ * @return 0 for success or negative in case of error.
+ *
+ */
+int lwm2m_engine_send_cb(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t path_list_size,
+		     coap_reply_t *reply_cb);
```

Comment:
interface design nitpick: using `coap_reply_t` kind of leaks implementation details. Might be worth defining a lwm2m callback for this (like `lwm2m_send_cb_t`)

Reply:
I was thinking about re-encapsulate the callback to be lwm2m related and not coap's.
Do you thing we shall define specific lwm2m code (in callback success/fail) or use the COAP(CODE_CHANGED, etc...) one ,so we still need to import coap.h in that last case

Reply:
I have no clear idea yet what would be best. There are success, failure and timeout cases.
Only in failure case you might want to inspect the coap response and would like to know the status code (and maybe even more like options). You made the argument in https://github.com/zephyrproject-rtos/zephyr/issues/52328 that timeout essential means failure for send, which I can also subscribe to. So one option would be to really boil it down to just a boolean, success or failure. 
On the other hand I can imagine use cases were one would like to know if is a hard failure (non-retryable, e.g. rejection by server) or soft failure (retryable, e.g. a timeout). So a result code (int or enum) could come in handy (e.g. success, failure, timeout). And if one really needs it, there could be some additional codes that represent certain COAP response codes. An alternative to bloating the result code with coap specific could be to pass in a context to the callback, that gives access to the coap_packet if there is one available.(that context could even contain a reference the message itself that could no be sent, for convenience of caller)

Reply:
I think I'm also leaning towards having a separate callback type `lwm2m_send_cb_t`

It could have just one enum as a parameter. I think we can minimize the cases to what you suggested: OK, rejected, timeout.

And I think this must be first handled inside engine, after all, the timeout is something that the LwM2M engine must react to.

CoAP return code ACCEPTED would be translated to OK/ACCEPTED/SUCCESS. Any other CoAP code would be translated as rejected by the server.
Timeout should cause LwM2M engine to drop into trying to re-register with the server. But timeout should be notified into this callback as well.

Oh.. and another thing, I wan't to get rid of those `lwm2m_engine_something` prefixes, so this would be:

```c
typedef void (*lwm2m_send_cb_t)(enum lwm2m_send_status status);

int lwm2m_send(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t path_list_size,
		    lwm2m_send_cb_t reply_cb);
```

Reply:
Initial callback are unchanged:
`msg->reply_cb = do_send_reply_cb;
msg->message_timeout_cb = do_send_timeout_cb;`

In case of timeout or COAP return success or error, lwm2m_send_cb_t callback is called with the corresponding status
`enum lwm2m_send_status {
	LWM2M_SEND_STATUS_SUCCESS,
	LWM2M_SEND_STATUS_FAILURE,
	LWM2M_SEND_STATUS_TIMEOUT,
};`

In case of multiple send pending, callback is stored in send_status_cb and struct lwm2m_message is stored in void *user_data of struct coap_reply to be retrieve the associated callback.
```
if (reply_cb) {
		msg->reply->user_data = msg;
		msg->send_status_cb = (lwm2m_send_cb_t) reply_cb;
	}
```